### PR TITLE
add AQL bit handling functions (#13332)

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/13332

This PR add the following bit handling functionality to AQL:

- BIT_AND(array): and-combined result
- BIT_OR(array): or-combined result
- BIT_XOR(array): xor-combined result
- BIT_NEGATE(value, bits): bitwise negation of `value`, with a mask of bits` length
- BIT_TEST(value, index): test if bit at position `index` is set in `value` (indexes are 0-based)
- BIT_POPCOUNT(value): return number of bits set in `value`
- BIT_SHIFT_LEFT(value, shift, bits): bitwise shift-left of `value` by `shift` bits, with a mask of `bits` length
- BIT_SHIFT_RIGHT(value, shift, bits): bitwise shift-right of `value` by `shift` bits, with a mask of `bits` length
- BIT_CONSTRUCT(array): construct a number with bits set at the positions given in the array
- BIT_DECONSTRUCT(value): deconstruct a number into an array with the set bits
- BIT_TO_STRING(value): create a bitstring representation from numeric `value`
- BIT_FROM_STRING(value): parse a bitstring representation into a number

`BIT_AND`, `BIT_OR` and `BIT_XOR` are also available as aggregate functions for usage inside COLLECT AGGREGATE.

All above bit operations support unsigned integer values with up to 32 bits. Using values outside the supported range will make any of these bit functions return `null` and register a warning.

The PR also adds binary (base 2) and hexadecimal (base 16) integer literals to AQL. 
These literals can be used where regular (base 10) integer literal can used. 
The prefix for binary integer literals is `0b`, e.g. `0b10101110`.
The prefix for hexadecimal integer literals i `0x`, e.g. `0xabcdef02`.

Binary and hexadecimal integer literals can only be used for unsigned integers.
The maximum supported value is `(2 ^ 32) - 1`, i.e. `0xffffffff` (hexadecimal) or `0b11111111111111111111111111111111` (binary).

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/607

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**

Link to Jenkins PR run:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13634/

### CODE REVIEW

**Comment by jsteemann** [comment]
Tests blue

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -6547,6 +6547,354 @@ AqlValue Functions::Pi(ExpressionContext*, transaction::Methods* trx,
...
+                                VPackFunctionParameters const& parameters) {
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  if (value.isNumber()) {
+    VPackSlice v = value.slice();
+    auto result = bitOperationValue<uint64_t>(v);
+    if (result.has_value()) {
+      uint64_t x = result.value();
+      // the following code fragment is from https://en.wikipedia.org/wiki/Hamming_weight#Efficient_implementation
+      // TODO: we could optimize this to use hardware-accelerated popcnt if it exists, or gcc's __builtin_popcntll
```
Comment (gnusi) [inline arangod/Aql/Functions.cpp:6756]: FYI
iresearch has `popcnt` implemented using compiler intrinsics (`utils/math_utils.hpp`), you can use it like
```
irs::math_utils::math_traits<decltype(x)>::pop(x)
```

Comment (jsteemann) [reply arangod/Aql/Functions.cpp:6756]: Ok, will do this in devel. Thanks for the info!

Comment (jsteemann) [reply arangod/Aql/Functions.cpp:6756]: Done in devel. Leaving the 3.7 as it is (also covered by tests, so results are also correct for 3.7).

File: arangod/Aql/Aggregator.cpp [hunk 2]
```diff
@@ -829,6 +830,93 @@ struct AggregatorCountDistinctStep2 final : public AggregatorCountDistinct {
...
+  AqlValue stealValue() override final {
+    if (invalid || !invoked) {
+      reset();
+      return AqlValue(AqlValueHintNull());
+    }
+
+    uint64_t r = result;
+    reset();
+    return AqlValue(AqlValueHintUInt(r));
+  }
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Aggregator.cpp:898]: It is intentionally different:
in 3.7, the Aggregator API is:
- reset()
- reduce(value)
- stealValue()
and stealValue() needs to call reset() internally.

in devel, the Aggregator API is:
- reset()
- reduce(value)
- get() const
and the caller needs to call reset().

I think it behaves correctly in both versions, although the code is (intentionally) different.

[the remark this answers is not in the document]
Comment (goedderz) [reply arangod/Aql/Aggregator.cpp:898]: Ah, I see. Thanks for the explanation!

**Review by fceller — APPROVED** [review]

**Pull request merged**